Building with Google Cloud Build (#496)
ordina-jworks/ordina-jworks.github.io: _posts/2019-03-28-Building-With-Google-Cloud-Build.md

---
layout: post
authors: [tom_verelst]
title: 'Building with Google Cloud Build'
image: /img/2019-03-28-cloudbuild/cloudbuild.png
tags: [GCP,Docker,DevOps,Git,Kubernetes]
category: Cloud
comments: false
---

In this post, 

> I will give a quick overview on what is possible with Google Cloud Build.

Reviewer: Gebruiken we normaal gezien niet de 'wij'-vorm in plaats van de 'ik'-vorm?

> Google Cloud Build is a fully managed solution to building containers or other artifacts.

Reviewer: ```suggestion
Google Cloud Build is a fully managed solution for building containers or other artifacts.
```

> It integrates with Google Storage, Cloud Source Repositories, GitHub or BitBucket.

Reviewer: ```suggestion
It can integrate with Google Storage, Cloud Source Repositories, GitHub and BitBucket.
```

Here is an example:

```
steps:
  # Test Helm templates
  - name: 'quay.io/helmpack/chart-testing:v2.2.0'
    id: 'Helm Lint'
    args: ['ct', 'lint', '--all', '--chart-dirs', '/workspace/helm', '--validate-maintainers=false']

  # Build image
  - name: 'gcr.io/cloud-builders/docker'
    id: 'Building image'
    args: ['build', '-t', 'eu.gcr.io/$PROJECT_ID/cloud-build-demo:$COMMIT_SHA', '.']

  # Create custom image tag and write to file /workspace/_TAG
  - name: 'ubuntu'
    id: 'Setup'
    args: ['bash', '-c', "echo `echo $BRANCH_NAME | sed 's,/,-,g' | awk '{print tolower($0)}'`_$(date -u +%Y%m%dT%H%M)_$SHORT_SHA > _TAG; echo $(cat _TAG)"]

  # Tag image with custom tag
  - name: 'gcr.io/cloud-builders/docker'
    id: 'Tagging image'
    entrypoint: '/bin/bash'
    args: ['-c', "docker tag eu.gcr.io/$PROJECT_ID/cloud-build-demo:$COMMIT_SHA eu.gcr.io/$PROJECT_ID/ms-map-report:$(cat _TAG)"]
images: ['eu.gcr.io/$PROJECT_ID/cloud-build-demo']
timeout: 15m
options:
  machineType: 'N1_HIGHCPU_8'
```

You are free to use any image that you like.
Cloud Build already [provides a set of base images (called Cloud Builders)](https://cloud.google.com/cloud-build/docs/cloud-builders),

> including images for Maven, Git, Docker, Bazel, npm, gcloud, kubectl and so on. 

Reviewer: ```suggestion
including images for Maven, Git, Docker, Bazel, npm, gcloud, kubectl, etc. 
```

> You can also customise some build options like the timeout time of the build,

Reviewer: ```suggestion
You can also customise some build options like the timeout of the build,
```

or on which kind of node the build runs.
Pricing is done based on the amount of build minutes. 
However, 
if you use the default node, 
the first 120 build minutes are free every day!

If the build finishes successfully,
Cloud Build will automatically upload the built images to the container registry.
This is based on the images defined in the `images` array.

> Data usually needs be shared in between steps.

Reviewer: ```suggestion
Data usually needs be shared between steps.
```

You might want to download dependencies in one step,
and build your artifact in another step,

> or run tests in a separate step.

Reviewer: You might want to download dependencies in step 1,
build your artifact in step 2 and run tests in step 3.

Google has provided a simple solution for this.
Each build step has access to the `/workspace` folder, 

> which is mounted on each step container.

Reviewer: ```suggestion
which is mounted on the container of each step.
```

Each build has access to its own workspace folder,
which is deleted automatically after the build finishes.

In the above example, 
a custom Docker tag is created and saved to the `/workspace/_TAG` file,
and then read from again in the next step.

To start the build you can use the `gcloud builds submit` command,
or create an automatic trigger on the Google Cloud console that triggers the build on new commits in the Git repository.
After adding a trigger, 

> you can also trigger the build manually on the Google Cloud console.

Reviewer: ```suggestion
you can also trigger the build manually in the Google Cloud console.
```

```
# Configure Docker
$ gcloud components install docker-credential-gcr
$ docker-credential-gcr configure-docker
```

Then install the `cloud-build-local` tool:

```
$ gcloud components install cloud-build-local
```

Now you can use the tool to test your build pipeline locally!

To build locally, run the following command:

```
$ cloud-build-local --config=[CONFIG FILE] \
  --dryrun=false \
  --push \
  [SOURCE_CODE]
```

* `CONFIG FILE` is your Cloud Build YAML config file
* `SOURCE_CODE` is the path to your source code
* `--dryrun=false` will cause your build to actually run. 
This is `true` by default and you must enable this explicitly to cause the containers to execute.
* `--push` will cause the built images defined in `images` to be pushed to the registry.

If you use some of the default substitions like `$COMMIT_SHA` in your build,
you must pass these in with the `--substitions` flag in key=value pairs,
separated by commas.
Example:

 ```
 $ cloud-build-local --config=cloud-build.yaml \
   --dryrun=false \
   --substitutions COMMIT_SHA=$(git rev-parse HEAD),BRANCH_NAME=$(git rev-parse  --abbrev-ref HEAD) \ 
   /path/to/source
 ```
 

> Cloud build stores intermediary artifacts in the workspace folder.

Reviewer: ```suggestion
Cloud Build stores intermediary artifacts in the workspace folder.
```

    ]
  },
  "logsBucket": "gs://199957373521.cloudbuild-logs.googleusercontent.com",
  "sourceProvenance": {
    "resolvedRepoSource": {
      "projectId": "mateco-map",
      "repoName": "bitbucket_matecocloud_myapp",
      "commitSha": "d76cce6d732e6edc01e65a547997caf107411468"
    }
  },
  "buildTriggerId": "9bd093c7-9de4-4eae-bfea-ce8e46afafa8",
  "options": {
    "substitutionOption": "ALLOW_LOOSE",
    "logging": "LEGACY"
  },
  "logUrl": "https://console.cloud.google.com/gcr/builds/a0e322f2-5c8d-4e56-a2b5-05cc18a350af?project=199957373521",
  "substitutions": {
    "_MOD_BRANCH_NAME": "$_tmpvar"
  },
  "tags": [
    "event-f2d96d7b-22f5-41d7-9ded-a98a2a6f43ca",
    "trigger-9bd093c7-9de4-4eae-bfea-ce8e46afafa8"
  ],
  "timing": {
    "BUILD": {
      "startTime": "2019-03-23T15:01:25.421114358Z",
      "endTime": "2019-03-23T15:15:58.377209942Z"
    },
    "FETCHSOURCE": {
      "startTime": "2019-03-23T15:01:20.519103589Z",
      "endTime": "2019-03-23T15:01:25.368505523Z"
    },
    "PUSH": {
      "startTime": "2019-03-23T15:15:58.377226850Z",
      "endTime": "2019-03-23T15:16:01.909032379Z"
    }
  }
}
```

> Note that Cloud Build does not publish events in between steps, 

Reviewer: ```suggestion
Note that Cloud Build does not publish events between steps, 
```

// Main function called by Cloud Functions.
module.exports.cloudBuildSlack = (event, callback) => {
    const build = eventToBuild(event.data.data);
    WEBHOOK.send(createSlackMessage(build), callback);
};
    
const createSlackMessage = (build) => {
    const app = getApplicationName(build);
    const branch = build.source.repoSource.branchName;
    const subject = createSubject(build);
    const tag = getImagetag(build);

    return {
        attachments: [{
            fallback: `${subject} - ${app} - ${branch} - <${build.logUrl}|Logs>`,
            title: subject,
            title_link: build.logUrl,
            fields: getFields(app, branch, tag),
            color: getMessageColor(build)
        }],
        mrkdwn: true
    };
};

// eventToBuild transforms pubsub event message to a build object.
const eventToBuild = (data) => {
    return JSON.parse(new Buffer(data, 'base64').toString());
};

...more functions
    
```

## Conclusion

Cloud Build offers a simple solution and utilises the power of containers to offer a lot of possibilities.
A build pipeline is set up in a few minutes, and your Docker images are uploaded automatically!

> It saves you a lot of time and trouble of setting up build infrastructure, 

Reviewer: ```suggestion
It saves you a lot of time and trouble in setting up build infrastructure, 
```
